.github/workflows/self-comment-ci.yml

```diff
@@ -0,0 +1,265 @@
+  cancel-in-progress: true
+
+jobs:
+  get-pr-number:
+    runs-on: ubuntu-22.04
+    name: Get PR number
+    # For security: only allow team members to run
+    if: ${{ github.event.issue.state == 'open' && contains(fromJSON('["ydshieh", "ArthurZucker", "zucchini-nlp", "qubvel", "molbap", "gante", "LysandreJik", "Cyrilvallez"]'), github.actor) && (startsWith(github.event.comment.body, 'run-slow') || startsWith(github.event.comment.body, 'run slow') || startsWith(github.event.comment.body, 'run_slow')) }}
+    outputs:
+      PR_NUMBER: ${{ steps.set_pr_number.outputs.PR_NUMBER }}
+    steps:
+      - name: Get PR number
+        shell: bash
+        run: |
+          if [[ "${{ github.event.issue.number }}" != "" && "${{ github.event.issue.pull_request }}" != "" ]]; then
+            echo "PR_NUMBER=${{ github.event.issue.number }}" >> $GITHUB_ENV
+          else
+            echo "PR_NUMBER=" >> $GITHUB_ENV
+          fi
+
+      - name: Check PR number
+        shell: bash
+        run: |
+          echo "${{ env.PR_NUMBER }}"
+
+      - name: Set PR number
+        id: set_pr_number
+        run: echo "PR_NUMBER=${{ env.PR_NUMBER }}" >> "$GITHUB_OUTPUT"
+
+  get-sha:
+    runs-on: ubuntu-22.04
+    needs: get-pr-number
+    if: ${{ needs.get-pr-number.outputs.PR_NUMBER != ''}}
+    outputs:
+      PR_HEAD_SHA: ${{ steps.get_sha.outputs.PR_HEAD_SHA }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: "0"
+          ref: "refs/pull/${{needs.get-pr-number.outputs.PR_NUMBER}}/merge"
+
+      - name: Get SHA (and verify timestamps against the issue comment date)
+        id: get_sha
+        env:
+          PR_NUMBER: ${{ needs.get-pr-number.outputs.PR_NUMBER }}
+          COMMENT_DATE: ${{ github.event.comment.created_at }}
+        run: |
+            git fetch origin refs/pull/$PR_NUMBER/head:refs/remotes/pull/$PR_NUMBER/head
+            git checkout refs/remotes/pull/$PR_NUMBER/head
+            echo "PR_HEAD_SHA: $(git log -1 --format=%H)"
+            echo "PR_HEAD_SHA=$(git log -1 --format=%H)" >> "$GITHUB_OUTPUT"
+            git fetch origin refs/pull/$PR_NUMBER/merge:refs/remotes/pull/$PR_NUMBER/merge
+            git checkout refs/remotes/pull/$PR_NUMBER/merge
+            PR_MERGE_COMMIT_TIMESTAMP=$(git log -1 --date=unix --format=%cd)
+            echo "PR_MERGE_COMMIT_TIMESTAMP: $PR_MERGE_COMMIT_TIMESTAMP"
+            COMMENT_TIMESTAMP=$(date -d "${COMMENT_DATE}" +"%s")
+            echo "PR_HEAD_SHA: $COMMENT_DATE"
+            echo "COMMENT_TIMESTAMP: $COMMENT_TIMESTAMP"
+            if [ $COMMENT_TIMESTAMP -le $PR_MERGE_COMMIT_TIMESTAMP ]; then
+              echo "Last commit on the pull request is newer than the issue comment triggering this run! Abort!";
+              exit -1;
+            fi
```

Comment:
new for security

Reply:
Currently I am using the merge commit.

Merge commit might change if there is a push or merge to the `main` branch. So it would be better to use the head commit of the PR instead.

Reply:
btw, not sure it's 100% safe, the commit date might be manually specified to be in the past 🥲

Reply:
however, this approach with comment message should be safer than triggering with contributor's commit message, because we control it

Reply:
Hi @qubvel This PR doesn't use users' input for date calculation. The merge commit `refs/remotes/pull/$PR_NUMBER/merge` is prepared by github itself.

But even if it's the date from the user (i.e. the actual commit from the contributor), would that date be manipulated?

Reply:
> This PR doesn't use users' input for date calculation. The merge commit refs/remotes/pull/$PR_NUMBER/merge is prepared by github itself.

Looks good then, thanks for clarification!

> But even if it's the date from the user (i.e. the actual commit from the contributor), would that date be manipulated?

I hope no! But it was an initial concern to disable the feature, wasn't it?

Reply:
> I hope no! But security issue was an initial concern to disable the feature, wasn't it?

Yes. At that time, there is no check on date at all. But now the concern is about if the commit date could be manipulated.
Anyway, if we use merge commit as mentioned above, it's fine :-)

Reply:
Hmm... if merge commit is created when CI is started wouldn't it be always later than the comment?

-> user commit
-> message to trigger CI (COMMENT_TIMESTAMP)
-> CI triggered
-> merge commit created (PR_MERGE_COMMIT_TIMESTAMP)

Reply:
Sorry, I should be more clear. The merge commit is prepared by github when a commit is pushed to the PR branch or the target branch of the PR (usually `main`) is updated.

So the flow is

-> user commit (pushed)
-> merge commit created (PR_MERGE_COMMIT_TIMESTAMP)
-> message to trigger CI (COMMENT_TIMESTAMP)
-> CI triggered

Reply:
Ok, sounds good then!